File: WINGS_AI_Bootcamp/Python/SortPolyEq.py
import re

def sort_polynomial(expr):
    terms = re.findall(r'[-+]?\s*\d*x(?:\^\d+)?|[-+]?\s*\d+', expr.replace(' ', ''))
    def key(term):
        if 'x^' in term:
            return -int(term.split('^')[1])
        elif 'x' in term:
            return -1
        else:
            return 0
    sorted_terms = sorted(terms, key=key)
    return ' '.join(sorted_terms).replace('+-', '- ')

expr = "3x^2 + 5x^4 - 2x + 7"
print(sort_polynomial(expr))

File: WINGS_AI_Bootcamp/Elixir/SortPolyEq.ex
defmodule Poly do
  def sort_poly(expr) do
    terms = Regex.scan(~r/[\+\-]?\s*\d*x(\^\d+)?|[\+\-]?\s*\d+/, expr) |> Enum.map(&hd/1)
    sorted = Enum.sort_by(terms, fn t ->
      cond do
        String.contains?(t, "^") ->
          [_x, pow] = String.split(t, "^")
          -String.to_integer(pow)
        String.contains?(t, "x") -> -1
        true -> 0
      end
    end)
    IO.puts Enum.join(sorted, " ")
  end
end

Poly.sort_poly("3x^2 + 5x^4 - 2x + 7")

